File: daemon/internal/evaluation/comparison_test.go
package evaluation

import (
	"context"
	"testing"
)

func TestCreateComparisonReportsDriftAndLimitations(t *testing.T) {
	ctx := context.Background()
	store := newMemoryStore()
	manager := NewManager(Dependencies{EnvironmentScope: "test", Store: store, Clock: fixedClock})
	candidate := ReplayCandidate{
		CandidateID:       "candidate_drift",
		CandidateKind:     CandidateKindCuratedWork,
		DisplayName:       "Runtime drift",
		SourceKind:        SourceKindRun,
		SourceID:          "run_base",
		EnvironmentScope:  "test",
		ReadinessStatus:   ReadinessFullyReplayable,
		DefaultReplayMode: ReplayModeNonLive,
		SourceRefs:        []SourceRef{{Kind: SourceKindRun, ID: "run_base", Route: "/v1/runs/run_base"}},
		ExpectedComparison: PlaneSummaries{
			Runtime:     "baseline runtime completed",
			Policy:      "baseline policy allowed",
			Integration: "baseline integration stable",
			Delivery:    "baseline delivery completed",
			Evidence:    "baseline evidence complete",
		},
	}
	if err := manager.UpsertReplayCandidate(ctx, candidate); err != nil {
		t.Fatalf("UpsertReplayCandidate returned error: %v", err)
	}
	attempt, err := manager.CreateReplayAttempt(ctx, candidate.CandidateID, CreateReplayAttemptInput{})
	if err != nil {
		t.Fatalf("CreateReplayAttempt returned error: %v", err)
	}
	attempt.RuntimeSummary = "replay runtime changed"
	if err := store.UpsertReplayAttempt(ctx, attempt); err != nil {
		t.Fatalf("UpsertReplayAttempt returned error: %v", err)
	}

	comparison, err := manager.CreateComparison(ctx, attempt.AttemptID, CreateComparisonInput{})
	if err != nil {
		t.Fatalf("CreateComparison returned error: %v", err)
	}
	if comparison.TerminalStatus != ComparisonDrifted {
		t.Fatalf("expected drifted comparison, got %s", comparison.TerminalStatus)
	}
	if len(comparison.DriftFindings) != 1 {
		t.Fatalf("expected one drift finding, got %+v", comparison.DriftFindings)
	}
	if comparison.DriftFindings[0].Plane != DriftPlaneRuntime {
		t.Fatalf("expected runtime drift, got %s", comparison.DriftFindings[0].Plane)
	}
}

File: rs/evaluation/tests/comparison_test.rs
//! Port of daemon/internal/evaluation/comparison_test.go.

mod common;

use std::sync::Arc;

use dope_evaluation::{
    CandidateKind, ComparisonTerminalStatus, CreateComparisonInput, CreateReplayAttemptInput,
    Dependencies, DriftPlane, Manager, PlaneSummaries, ReadinessStatus, ReplayMode, SourceKind,
    SourceRef,
};

use common::{MemoryStore, fixed_now};

#[tokio::test]
async fn create_comparison_reports_drift_and_limitations() {
    let store = Arc::new(MemoryStore::new());
    let manager = Manager::new(Dependencies {
        environment_scope: "test".to_string(),
        store: Some(store.clone()),
        clock: Some(Arc::new(fixed_now)),
        ..Default::default()
    });
    let candidate = dope_evaluation::ReplayCandidate {
        candidate_id: "candidate_drift".to_string(),
        candidate_kind: CandidateKind::CuratedWork,
        display_name: "Runtime drift".to_string(),
        source_kind: SourceKind::Run,
        source_id: "run_base".to_string(),
        environment_scope: "test".to_string(),
        readiness_status: ReadinessStatus::FullyReplayable,
        default_replay_mode: ReplayMode::NonLive,
        source_refs: vec![SourceRef {
            kind: SourceKind::Run,
            id: "run_base".to_string(),
            route: "/v1/runs/run_base".to_string(),
        }],
        expected_comparison: Some(PlaneSummaries {
            runtime: "baseline runtime completed".to_string(),
            policy: "baseline policy allowed".to_string(),
            integration: "baseline integration stable".to_string(),
            delivery: "baseline delivery completed".to_string(),
            evidence: "baseline evidence complete".to_string(),
        }),
        ..Default::default()
    };
    manager
        .upsert_replay_candidate(candidate)
        .expect("UpsertReplayCandidate returned error");
    let attempt = manager
        .create_replay_attempt("candidate_drift", CreateReplayAttemptInput::default())
        .await
        .expect("CreateReplayAttempt returned error");
    let mut attempt = attempt;
    attempt.runtime_summary = "replay runtime changed".to_string();
    store.insert_attempt(attempt.clone());

    let comparison = manager
        .create_comparison(&attempt.attempt_id, CreateComparisonInput::default())
        .expect("CreateComparison returned error");
    assert_eq!(
        comparison.terminal_status,
        ComparisonTerminalStatus::Drifted,
        "expected drifted comparison"
    );
    assert_eq!(comparison.drift_findings.len(), 1);
    assert_eq!(comparison.drift_findings[0].plane, DriftPlane::Runtime);
}
